### ISSUE TITLE
Using Random forest behavioral classifiers from OSF repository 

### ISSUE THREAD

**Issue by wallawhitecoat** [sgoldenlab/simba#367]
Hi, I'm trying to run resident intruder classifiers on one imported video, ten mins long, 60fps, and over 36,000 frames. I've been able to successfully import my H5 tracking data from SLEAP. 
My skeleton is SLEAP based, 11 body parts per animal. 
![User Defined Skeleton](https://github.com/sgoldenlab/simba/assets/169295221/cddf468e-5682-47dc-b4b4-cb3b2d21e29f)

I have four questions. 
**1. Is it possible to run the following behavioral classifiers (anogenital_sniff, attack, lateral_threat, pursuit, & tail_rattle) I found on the Random forest behavioral classifiers OSF repository on my data if the body-part config is user defined? If yes, am I correct in assuming that I can use PSEUDO-Labelling for this? **
https://osf.io/3mc7g/

If the answer to this is no, I am fine with remedying this by deleting unnecessary nodes. 

**2. Is this the feature extractor (from SimBA Github) that I should be using, it seems to freeze once I use it?** 
[feature_extractor_user_defined.zip]
(https://github.com/user-attachments/files/15779886/feature_extractor_user_defined.zip)
![image](https://github.com/sgoldenlab/simba/assets/169295221/f6d16120-fa0a-4825-9f61-5e810945a11f)

**3. Downloading CSV files from SLEAP puts both tracks (for resident and intruder) in a single column, it seems that to train the model, SimBA wants each node annotated in individual columns (example "Resident_forelegL1_1_y"). I would like to know how to remedy this.** 
Example Below: How SLEAP CSV files are exported and labeled
![How SLEAP CSV files are exported and labeled](https://github.com/sgoldenlab/simba/assets/169295221/a725a0ec-11ae-4dec-87e2-41f64c5a3ada)

4. Should I be using this meta data file? 
[BtWGaNP_meta.csv](https://github.com/user-attachments/files/15779964/BtWGaNP_meta.csv)
https://github.com/sgoldenlab/simba/blob/master/misc/BtWGaNP_meta.csv

Here is my Outlier correction settings in case it helps:
![image](https://github.com/sgoldenlab/simba/assets/169295221/7eb06a2c-0fb8-488f-b954-ad5783077565)

Any help is appreciated, thank you! 

Windows 10, version 22H2
Python Version [3.10.14]
using miniconda3 Version [24.4.0]

**Comment by sronilsson** [comment]
Hi @[wallawhitecoat](https://github.com/wallawhitecoat)!


1. To use the models on OSF, you need to track the same body-parts that were used to create the model. This ensures that SimBA computes the same features that the models expects. If you have different body-parts configurations, SimBA will compute different features and a different number of features, and the models will be confused what to do with the additional / missing features and throw you an error. To use these models, use this setting when creating your project:

￼


2. One possibility that it “freezes”, is that the example script is expecting data columns that don’t exist in your data, and it errors out. If it errors out in a way that I haven’t yet been able to anticipate, no errors will be printed in the main SimBA interface. Instead, an error message from the standard python library will be printed in the main terminal which you used to launch SimBA. If you look in the terminal that you used to launch SimBA, do you see any error msg printed? 


3. On the difference between the SLEAP data, and the format in SimBA where the SLEAP data is transposed. When you import your data into SimBA, the data lands in the `project_folder/csv/input_csv` within your SimBA project. After you perform outlier correction (or skip to perform outlier correction), the corrected data is copied to the `project_folder/csv/outlier_corrected_location` `and project_folder/csv/outlier_corrected_movement_location` directories of your SimBA project with the additional corrected headers appended. If you want the transposed data, you can look in those folder.

4. Yes you can use [THIS](https://github.com/sgoldenlab/simba/blob/master/misc/BtWGaNP_meta.csv) hyper parameter meta file to create models to start. However, you may have to play with it, in particular the under sampling ratios, to get a model that performs best in your setup.

**Comment by sronilsson** [comment]
No sorry - the tail end classifier is legacy. We had some trouble with this - it was very difficult to get tail end tracked reliably even after extensive training and labelling - you may have more luck though. It was often confused with the bedding material and we saw a lot of ID swappes with the tail end of the two animals beeing confused with each other.

**Comment by wallawhitecoat** [comment]
Yes, I did experience that even after training close to 2000 frames. Could we use the tail_rattle classifier if the configuration looks like this? 
![SLEAP settings](https://github.com/sgoldenlab/simba/assets/169295221/77a6aac4-2a7b-4909-b16e-d16502846988)

Also, the reason I ask is because in this youtube video I don't see a tail end but tail_rattle is still being quantified. 
https://www.youtube.com/watch?v=bqWteWIxzGM

**Comment by sronilsson** [comment]
Yes - that should work. But - just for full disclosure - I created the tail rattle classifier in Seattle around 2019. Everyone who worked on this back then, including maintaining the OSF repository, have other interests and jobs today. I don't remember exactly which features the tail end classifier excepts, and no-one is available for me to ask. If you hit errors, I can start to dig through for old data and figure it out though and we could solve it together. 

A reason why it is not visualized is that although the classifications may be OKish: for that video we wanted to visualize the hull polygonal bounding box of the animals in white. We probably omitted the tail end to get that white polygon box thing around the animal to look neater.

**Comment by wallawhitecoat** [comment]
Ok, took your advice, back after re-training and labeling with this skeleton. 

1. So, I'm able to run the feature extraction, but is a file supposed to be inserted in the project_folder ->csv->features_extracted? Because I do not see anything. 
![Feature Extraction](https://github.com/sgoldenlab/simba/assets/169295221/537ecb1f-7c80-4c13-bf6a-0049165e9df9)
![image](https://github.com/sgoldenlab/simba/assets/169295221/e38ec6b7-e810-437d-a780-4f7c52d5b201)

2. Additionally, I skipped the outlier correction, and the SLEAP data was successfully transposed into outlier_corrected_movement_location. I guess I mostly don't exactly understand what completing an outlier correction will do for my data, or what settings I should run it on. 

3. My last question resolves around how I should label. Do I just label all frames using the regular "label behavior" method? Or if I'm using the sav. classifiers should I use pseudo-labelling and/or advanced labeling?

**Comment by wallawhitecoat** [comment]
Additional note on Question 3 here, I noticed that if I attempt to **Validate on a Single Video** using the .sav files I get this error. Is that because I did not perform my feature extraction correctly? 
![image](https://github.com/sgoldenlab/simba/assets/169295221/69f760a2-d6b1-4e6c-9814-4b825555993d)
![image](https://github.com/sgoldenlab/simba/assets/169295221/b9dd954b-d9cd-4397-a927-eb5e53fbc6e1)

When I attempt to **run the machine model** with these settings I get this error message, which also notes to the features_extracted directory. 
![models](https://github.com/sgoldenlab/simba/assets/169295221/5f87f708-8663-4400-b2b4-2e97bdc4419c)
![image](https://github.com/sgoldenlab/simba/assets/169295221/5bf33899-867e-4cfa-b7b4-9371e29ba2c8)

**Comment by sronilsson** [comment]
Hi @wallawhitecoat - 

1) Yes, it appears to go astray at the point of features extraction, with no files being created. From the screenshot, it looks like you have ticked the box to run your own feature extraction script like in this image below is this correct? 

<img width="387" alt="image" src="https://github.com/sgoldenlab/simba/assets/34761092/032cad28-10db-4af9-8fc5-cb021b946f3c">

If you have ticked this box, could you share with me the .py file you have used in the `Script path` file browse box? From your screenshoot, it looks like the script is running appropriately, but the final file might be saved in the wrong location or not saved at all. 

2) For outlier correction, I've tried to wrote some explanations [HERE](https://simba-uw-tf-dev.readthedocs.io/en/latest/tutorials_rst/scenario_1.html#step-4-outlier-correction) and [HERE](https://github.com/sgoldenlab/simba/blob/master/docs/Scenario1.md#step-4-outlier-correction). In short, what it is trying to do, is to fix big pose-estimation inaccuracies based on user-defined (heuristic) rules. It may be that a body-part can't move N x the length of the animal in a single frame, or that a body-part can't be located distance N x the length of the animal from all the other body-parts of the animal. If it fails those rules, the body-parts are placed in their most recent reliable location. If you have very good pose-estimation tracking, you shouln't have to apply these rules though. 

3) I've reached out for some help to answer this one I think @goodwinnastacia can help

**Comment by wallawhitecoat** [comment]
Thank you! I was able to perform a feature extraction just using the "extract features" button and not with the user-defined settings. I was then able to use the .sav files after deleting the following features, the .sav files only recognize 490 features, not 498 features. I also had to add a prefix for the pose estimation locations (ex. "**track_1**_Ear_left_1_x")
![image](https://github.com/sgoldenlab/simba/assets/169295221/ed6ea175-b88e-414e-8c99-2665b27a7e68)

However, after running the machine model and generating the GANTT graphs and video, I found that the classifiers were too sensitive for some, and not sensitive enough for others. I've tried multiple discrimination thresholds, so far 0.1 is too low and quantifies behavior when it doesn't happen, and 0.7 is too high and doesn't quantify anything. My pose estimation tracking is pretty good, so I'm not exactly sure what the issue is, but thank you for the resource I will follow up with them. Thank you again for helping me troubleshoot! 
![labels v2001_final_image](https://github.com/sgoldenlab/simba/assets/169295221/1b62d2fa-97f9-4ca5-a4f9-6c33bb3ae9ab)
c
![image](https://github.com/sgoldenlab/simba/assets/169295221/0005ec49-fe8e-403b-b687-9387924d3ab1)

**Comment by sronilsson** [comment]
Hi @wallawhitecoat  - thanks for sharing - and thanks for figuring out the additional features that had to be removed - I wasn't sure if that would come up and it was in the back of my mind. 

For the predictions - could we confirm that it is not the tracking that has been disrupted somehow along the process? Can you run the the classification visualization in the menu below and share the video? 

<img width="977" alt="image" src="https://github.com/sgoldenlab/simba/assets/34761092/236063e4-7036-491d-b1d0-0d1a4a624369">

**Comment by goodwinnastacia** [comment]
> 3\. My last question resolves around how I should label. Do I just label all frames using the regular "label behavior" method? Or if I'm using the sav. classifiers should I use pseudo-labelling and/or advanced labeling?

Hi, I would always start by labeling a few full videos using the label behavior method just so that you have a good representation of what tail rattle ISN'T, since you have to teach the algorithm both what positive and negative frames look like. I would then create short clips with tail rattle present and label those with the label behavior method. Once you've got some videos labeled this way, you can take a look at using pseudolabeling. 

For the tail rattle classifier we have  posted, our tracking wasn't good enough for good behavioral classification because we only did tail base and tail end. Your tracking should be better!

**Comment by wallawhitecoat** [comment]
Of course, thank you for the advice, I performed the visualize quantifications (with the body-part visualization threshold set to 0.0), here are the results (no classifiers shown). It's a youtube link because the video was too large to post (even when zipped). Below you can see the settings I used and my rational.  
https://www.youtube.com/watch?v=THQ2TnCWfsw

Decided to try these settings for running the machine labelling because they are detailed ([in this paper](https://www.nature.com/articles/s41593-024-01649-9)) and I just set the tail_rattle classifier to 0.5 (even if it isn't good enough). Still not sure what to set my min. bout length to though. 
![image](https://github.com/sgoldenlab/simba/assets/169295221/e69298b5-16cd-432a-a2df-c1810928a269)
![image](https://github.com/sgoldenlab/simba/assets/169295221/7d054561-7f40-456b-aa4a-3e7dfac3f91c)

As for labelling full videos, the reason I wanted to use existing classifiers is because they had been recently published in that paper, so I had thought users would be able to apply them to their own data (I had found this information from this quote in the paper **"To perform supervised behavioral classification, users can download pre-made classifiers from our OSF repository, request classifiers from collaborators or create classifiers by annotating new videos in the scoring interface"**. 

How would the existing classifiers/.sav files fit into my training If I manually annotate all the behaviors on my own? Would a wise course of action be to just get machine-results using the existing classifiers, and then do pseudo labeling? 

Thank you again for your continued help on this :)

**Comment by goodwinnastacia** [comment]
So there are two potential ways to do this. 1) download our videos, track them in SLEAP using your 11 point model, and then append our ground truth annotations. I think this is advantageous because the extra tail point is going to help with your tail rattle classifications. 2) Download our targets inserted files off of OSF, add them to your project folder -> csvs -> targets inserted folder, and download the video info for those projects and add them to your own video info log. This will use our original DLC tracking for our videos with the less than ideal tail end tracking and could potentially make your classifications a bit worse at first. The advantages of this option are that you're diversifying your tracking so your models should ultimately be more robust, and you don't have to run more videos through pose estimation. Let me know which you'd like to move forward with and I can help you out with it. 

It does look like there's a glitch on OSF right now and our files aren't showing up, but I should have that fixed by the end of the weekend.

**Comment by wallawhitecoat** [comment]
I think the second method would work out well, not having a perfect tail_rattle classifier is not a deal breaker for my team. If we could have attack and anogenital sniff work well that would be satisfactory. Also, no worries thank you for fixing!

**Comment by goodwinnastacia** [comment]
Hi, to clarify - you currently have files with 16 total body parts and 498 features, correct? 

I put the hand-annotated training sets on OSF and you can find them through the filepaths shown in the screenshot. You should be able to copy these csv files into your targets inserted folder. After that,  you need to add the video names and associated info to your "video_info" log in the project_folder -> logs folder. This info can be found in the video_info folder in the folder: 16BP_498_Features_need_refinement_PREPRINT_CLASSIFIERS -> Video_Info.

Please let me know if you have any questions!

<img width="709" alt="Screenshot 2024-07-03 at 2 31 27 PM" src="https://github.com/sgoldenlab/simba/assets/59748563/f8e657c9-26b0-42aa-a8f2-5ae7cf0b78ca">

**Comment by wallawhitecoat** [comment]
Yes, they are 16 body parts. I had to manually remove 8 features from the features_extracted file in order to use the .sav classifiers, I assume those 8 had been added to SimBA after the .sav classifiers were put on the OSF (see the above excel spreadsheet to see which ones were removed. For what I did below, I did not remove those 8 features. 

After I added the annotations and the corresponding video_info file, I found I could Train multiple models (one for each saved setting) with the error below mentioned. I saved the settings as they were, this .csv file appeared in configs. 
![image](https://github.com/sgoldenlab/simba/assets/169295221/f3e67a90-c927-4fc9-b831-7b3c6fc765f4)
![image](https://github.com/sgoldenlab/simba/assets/169295221/ca8ac823-9762-4a1f-af43-b6f5ac31a76d)

Here is what SimBA output for me, it took about a half hour. 
![image](https://github.com/sgoldenlab/simba/assets/169295221/171fe07e-3ea9-4dae-a4d4-7916b2a9a00c)
![image](https://github.com/sgoldenlab/simba/assets/169295221/0872be4b-8375-4b8b-b260-65ce843d20d0)
![image](https://github.com/sgoldenlab/simba/assets/169295221/55a074fb-85e3-40e9-aed8-deb8cf48dd25)

Then I re-named the .sav file to just "Tail_rattle" and ran "Validate model on single video" . 
![image](https://github.com/sgoldenlab/simba/assets/169295221/04663b5b-8b86-4057-b758-ef50565238f3)

Based off of the graph, it didn't seem that the classifier was more accurate, so I created a video w/a GANTT chart **(Discrimination threshold- 0.4, Minimum bout length (MS)- 1)**. I think it is more accurate than my previous results, but still needs some work. I'll attach a video link. 

https://youtu.be/cdtEezm4i7c

Anyways, all this info is just to ask what the next step is. Additionally, is this something I should replicate for the other classifiers to improve their accuracy?  Thank you!

**Comment by goodwinnastacia** [comment]
<img width="1156" alt="image" src="https://github.com/sgoldenlab/simba/assets/59748563/bef1c671-34d5-4059-9d7f-bce43f77d1eb">

**Comment by goodwinnastacia** [comment]
Have your thresholding graphs always looked like that, or did they previously look more like the top two graphs in the flowchart I just sent? Usually when we see a threshold plot like this without a low baseline, there's some error in the annotations

**Comment by wallawhitecoat** [comment]
The thresholding graphs have looked like this so far, what should I do to fix this? I also wanted to know if I need the corresponding videos for the targets_inserted files. So far, I have only run this on my 10min long video, not any on the OSF.

**Comment by goodwinnastacia** [comment]
Can you please send me one of your targets inserted files that you hand annotated, and then one from the training sets we provided? I want to make sure that no columns got scrambled. You can download the OSF vids if you want to double check the annotations or something like that, but you don't need to if you don't want to visualize anything.

**Comment by wallawhitecoat** [comment]
Sure, also I finished completely labeling my 10 min long video (with all of the classifiers that I wanted to use), and was able to successfully generate my first set of kind-of accurate classifiers (yay!) I'll attach a video, the accuracy is much improved for the tail rattle classifier, but still not where I want it to be. 
https://www.youtube.com/watch?v=AngsHVoKBs0

(NOTE: I had to rename the classifier from "Tail_rattle" to "tail_rattle" in the training sets for SimBA to recognize it)
The files were too large (even zip files) to put here on GitHub, so I uploaded them to the OSF Repository ([link](https://osf.io/yfcbp/)) 
training set file name ->CSDS02303.csv
mine -> labels.v2001.csv

I do think that the columns got scrambled. When I tried to run a training alongside the targets_inserted files you provided to make my tail_rattle classifier better, SimBA printed these errors and gave me this log result. I removed all of the other classifiers and tried to run it again, but it still didn't work. 
I think that some of the features in your targets_inserted files have a. different names for the same feature or b. do not exist anymore/were replaced by other features
![image](https://github.com/user-attachments/assets/82028dd5-d21a-4b25-bf02-986f165d830c)
Log -> [missing_columns_20240716151532.csv](https://github.com/user-attachments/files/16257325/missing_columns_20240716151532.csv)

Let me know if there's any work-around on this, thank you!

**Comment by goodwinnastacia** [comment]
Is labels.v2001 your only video in the training set?

**Comment by wallawhitecoat** [comment]
Yes, it's the only one of my own that I have fully labelled in the targets_inserted folder.

**Comment by goodwinnastacia** [comment]
Okay. Would you mind taking a look at this video that I made on how to create classifiers? They way you're doing it is valid, with one project for all of your classifiers. Another way to do it, which I prefer because it's ultimately much less work, is to make a project for each individual classifier. That way you don't have to annotate every video for every behavior. The annotations that I put on OSF are designed for the one project, one classifier way of doing things. It won't be hard to carry the work you've already done over into that project format, just some copy and pasting. https://www.youtube.com/watch?v=qqwm6hMuKI4

**Comment by wallawhitecoat** [comment]
Hi, sorry for the radio silence, we were labeling our own videos. 

I tried just making one project for the tail_rattle classifier using your target_inserted files. Our targets_inserted files are 10 minutes long, 60fps, and the ones I imported from your data are much shorter. Would that effect the model? I only ask this because I trained it recently using both, and it stops classifying behavior at around 5 minutes into our videos. Here is an interactive probability plot from a video with lots of tail rattling and the discrimination threshold set at 0.5, minimum bout length of 1ms. 
![image](https://github.com/user-attachments/assets/1bb7314c-0633-4e44-b26b-b5ed58371770)

**Comment by goodwinnastacia** [comment]
It shouldn't have any impacts on the model performance. Do all of your videos look like this? Did you train your model on full videos, or only the first half? It looks to me like the classifier is still doing what it should, there's just something different about the tail rattle in the second half of your videos that the classifier isn't as sure about. You can fix it by scoring the second half of any videos like this (trim them in the video trimmer then score the resulting video) and adding it to your training set.
